src/layers/core/hexagon-layer/hexagon-layer.js

```diff
@@ -0,0 +1,221 @@
+   * HexagonLayer is a variation of grid layer, it is intended to render
+   * hexagon tessellations. It supports elevation, lighting as well
+   *
+   * @class
+   * @param {object} props
+   * @param {number} props.dotRadius - hexagon radius multiplier
+   * @param {number} props.elevation - hexagon elevation uniform, 1 unit roughly equals 100 meters
+   * @param {boolean} props.enable3d - if set to false, all hexagons will be flat
+   * @param {array} props.hexagonVertices - primitive hexagon vertices as [[lon, lat]]
+   * @param {object} props.invisibleColor - hexagon invisible color
+   * @param {function} props.getCentroid - hexagon centroid should be formatted as [lon, lat]
+   * @param {function} props.getColor -  hexagon color should be formatted as [255, 255, 255]
+   * @param {function} props.alpha -  hexagon opacity should be from 0 - 255
+   * @param {function} props.getElevation - hexagon elevation 1 unit approximate to 100 meters
+   *
+   */
+  constructor(props) {
+    assert(props.hexagonVertices, 'hexagonVertices must be supplied');
+    super(Object.assign({}, defaultProps, props));
+  }
+
+  /**
+   * DeckGL calls initializeState when GL context is available
+   * Essentially a deferred constructor
+   */
+  initializeState() {
+    const {gl} = this.context;
+    this.setState({model: this.getModel(gl)});
+
+    const {attributeManager} = this.state;
+    attributeManager.addInstanced({
+      instancePositions: {size: 3, update: this.calculateInstancePositions},
+      instanceColors: {size: 4, update: this.calculateInstanceColors}
+    });
+
+    this.updateRadiusAngle();
+  }
+
+  updateState({props, oldProps, changeFlags: {dataChanged, viewportChanged}} = {}) {
+    const {model, attributeManager} = this.state;
+
+    if (dataChanged) {
+      attributeManager.invalidateAll();
+    }
+
+    // Update the positions in the model if they've changes
+    const verticesChanged =
+      !positionsAreEqual(oldProps.hexagonVertices, props.hexagonVertices);
+
+    if (model && (verticesChanged || viewportChanged)) {
+      this.updateRadiusAngle();
+    }
+
+    this.updateUniforms();
+  }
+
+  updateRadiusAngle() {
+    const {hexagonVertices: vertices, dotRadius} = this.props;
+
+    const vertex0 = vertices[0];
+    const vertex3 = vertices[3];
+
+    // transform to space coordinates
+    const spaceCoord0 = this.projectFlat(vertex0);
+    const spaceCoord3 = this.projectFlat(vertex3);
+
+    // distance between two close centroids
+    const dx = spaceCoord0[0] - spaceCoord3[0];
+    const dy = spaceCoord0[1] - spaceCoord3[1];
+    const dxy = Math.sqrt(dx * dx + dy * dy);
+
+    this.setUniforms({
+      // Calculate angle that the perpendicular hexagon vertex axis is tilted
+      angle: Math.acos(dx / dxy) * -Math.sign(dy) + Math.PI / 2,
+      // Allow user to fine tune radius
+      radius: dxy / 2 * Math.min(1, dotRadius)
+    });
+  }
+
+  getCylinderGeometry(radius) {
+    return new CylinderGeometry({
```

Comment:
For performance with large data sets in 2D, it could be valuable to support an `extruded` prop, and if false, generate a simpler non-extruded hexagon primitive. For now it would need a simple [0, 1, 0] normal to work with the lighting code in the shader, until we make that optional as well.